File: ts/test/examples.test.ts
import { test, expect } from "bun:test";
import { load } from "../src/index.ts";

const AND = new URL("../../circuits/and.circ", import.meta.url).pathname;
const FULL_ADDER = new URL("../../circuits/full_adder.circ", import.meta.url).pathname;

test("and gate: exhaustive a & b", async () => {
  const dut = await load(AND);
  try {
    for (let a = 0; a <= 1; a++) {
      for (let b = 0; b <= 1; b++) {
        await dut.set("a", a);
        await dut.set("b", b);
        await dut.settle();
        expect(await dut.getInt("out")).toBe(BigInt(a & b));
      }
    }
  } finally {
    await dut.close();
  }
});

test("full adder: sum and carry via eval, computed reference", async () => {
  const dut = await load(FULL_ADDER);
  try {
    for (let a = 0; a <= 1; a++) {
      for (let b = 0; b <= 1; b++) {
        for (let cin = 0; cin <= 1; cin++) {
          const r = await dut.eval({ a, b, cin }, ["sum", "cout"]);
          const total = a + b + cin;
          expect(r.sum.value).toBe(BigInt(total & 1));
          expect(r.cout.value).toBe(BigInt(total >> 1));
        }
      }
    }
  } finally {
    await dut.close();
  }
});

test("undefined input propagates to undefined output", async () => {
  const dut = await load(AND);
  try {
    await dut.setUndefined("a");
    await dut.set("b", 1);
    await dut.settle();
    expect(await dut.isUndefined("out")).toBe(true);
  } finally {
    await dut.close();
  }
});

File: zig/src/examples_test.zig
const std = @import("std");
const circ = @import("circ_test");
const build_options = @import("build_options");

fn loadCircuit(name: []const u8) !circ.Dut {
    var buf: [std.fs.max_path_bytes]u8 = undefined;
    const path = try std.fmt.bufPrint(&buf, "{s}/{s}", .{ build_options.circuits_dir, name });
    return circ.Dut.load(std.testing.allocator, path);
}

test "and gate: exhaustive a & b" {
    var dut = try loadCircuit("and.circ");
    defer dut.close();
    var a: u64 = 0;
    while (a <= 1) : (a += 1) {
        var b: u64 = 0;
        while (b <= 1) : (b += 1) {
            try dut.set("a", a);
            try dut.set("b", b);
            try dut.settle();
            try std.testing.expectEqual(a & b, try dut.getInt("out"));
        }
    }
}

test "full adder: sum and carry, computed reference" {
    var dut = try loadCircuit("full_adder.circ");
    defer dut.close();
    var a: u64 = 0;
    while (a <= 1) : (a += 1) {
        var b: u64 = 0;
        while (b <= 1) : (b += 1) {
            var cin: u64 = 0;
            while (cin <= 1) : (cin += 1) {
                try dut.set("a", a);
                try dut.set("b", b);
                try dut.set("cin", cin);
                try dut.settle();
                const total = a + b + cin;
                try std.testing.expectEqual(total & 1, try dut.getInt("sum"));
                try std.testing.expectEqual(total >> 1, try dut.getInt("cout"));
            }
        }
    }
}

test "undefined input propagates to undefined output" {
    var dut = try loadCircuit("and.circ");
    defer dut.close();
    try dut.setUndefined("a");
    try dut.set("b", 1);
    try dut.settle();
    try std.testing.expect(try dut.isUndefined("out"));
}
